[WIP] (PCP-489) Refactor and improve onMessage handling

parisiale opened the pull request:
Still WORK IN PROGRESS as:
 - [x] I must update the integration tests;-
 - [x] I may add new tests to core_test (PCP-523);
 - [ ] there are a number of TODOs to be resolved;
 - [ ] I will try to factor out PCP-385 changes and split the commit in
   different pieces.

Checking the validity of incoming messages only once, meaning:
 - ignore messages other than associate_request before Association is
   complete;
 - authentication;
 - authorization;
 - TTL check.

Removing the transitions map that keept track of the state of the
association during the message double dispatch mechanism in favour of
making explicit the processing workflow.

Avoiding to update the 'connections' map for each processing message. In
practice, the onMessage handler will update the map only if the
Connection instance being processed is actually updated; that happens
only in case of a successful associate_request.

In case of invalid (as above) associate_request, an unsuccessful
associate_response is sent back to the requester and the WebSocket
connection is closed. That's a behaviour change, as:
 - in case of authentication failure, the broker used to send a PCP
   error and close the WebSocket;
 - in case of authorization failure, the broker used to only close the
   WebSocket.

parisiale commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -281,216 +252,319 @@
+            :debug {:type   :connection-association-failed
+                    :uri    requester-uri
+                    :reason reason-to-deny}
+            (i18n/trs "Invalid associate_request ({reason}); closing '{uri}' WebSocket"))
```
Missing single quotes for `reason`.

parisiale commented on test/utils/puppetlabs/pcp/testutils/client.clj:
```diff
@@ -60,9 +60,9 @@
                                                      (http/send ws :byte (message/encode association-request)))
                                             :byte  (fn [ws msg]
-                                                     (>!! message-chan (message/decode msg)))
+                                                     (put! message-chan (message/decode msg)))
```
The blocking '>!' was effectively block those tests that make the broker reply with a sequence of unsuccessful 'associate_response' / WebSocket close.

MikaelSmith commented:
Success!

MikaelSmith commented on src/puppetlabs/pcp/broker/connection.clj:
```diff
@@ -45,17 +46,22 @@
+     ;; http://docs.oracle.com/javase/7/docs/api/java/net/InetAddress.html#getHostAddress%28%29
+     :remote-address (try+
+                       (second
+                         (str/split (.. (websockets-client/remote-addr websocket) (toString)) #"/"))
```
What was the problem with the prior implementation using getSession?

parisiale replied:
Two things here. We have the 'getSession' - 'getRemoteAddress' exposed in `trapperkeeper-webserver-jetty9`, so we shouldn't be going directly to jetty.

As for the parsing, the intent here is to see if this makes sense and later move upstream; 'getRemoteAddress' returns an 'InetSocketAddress' that has a 'toString' method that ends up calling the 'InetAddress' parent 'toString' that, in turn, returns a string in the 'hostname / literal IP address' format; since Jetty does not resolve the hostname, the string becomes something like '/0:0:0:0:0:0:0:1:56824'.

Another way to solve this, perhaps more correct if we were using Java directly :trollface: , would be calling 'InetAddress' 'getAddress' directly as suggested in http://stackoverflow.com/questions/6932902/apache-mina-how-to-get-the-ip-from-a-connected-client

MikaelSmith commented on locales/messages.pot:
```diff
@@ -105,10 +102,11 @@ msgstr ""
 
+#. TODO(ale): is 'superceded' correct? 'superseded'?
 #: src/puppetlabs/pcp/broker/core.clj
 msgid "superceded"
```
superseded is more correct (though both are accepted in common use)

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -146,6 +109,7 @@
   (let [sender (:sender message)
         in-reply-to (:id message)
         response_data (->> {:id in-reply-to}
+                           ;; TODO(ale): consider removing s/validate; we create it, why validate?
```
I'd remove it. If we really want checking, you could make a helper function like
```
(s/defn make-ttl_expired-envelope :- p/TTLExpiredMessage
  [in-reply-to :- s/Str]
  {:id in-reply-to})
```

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -209,6 +177,7 @@
   (when (:destination_report message)
     (let [report {:id (:id message)
                   :targets targets}]
+      ;; TODO(ale): consider removing s/validate; we create it, why validate?
```
Same as above, I'd make a helper function
```
(s/defn make-destination-report :- p/DestinationReport
  [id targets]
  {:id id
   :targets targets})
```

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -281,216 +252,323 @@
+        (i18n/trs "Session already associated")))))
+
+(s/defn process-associate-request! :- (s/maybe Connection)
+  "Send an associate_response that will be successfull if:
```
Should be `successful`.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -281,216 +252,323 @@
+    - the requester `client_type` is not `server`;
+    - the specified WebSocket connection has not been associated previously.
+  If the request gets denied, the WebSocket connection will be closed and the
+  function returns nil.
```
If the connection is closed, should we remove it from the connection and uri maps?

parisiale replied:
The current and previous behaviour consists of updating the broker state with `remove-connection!` only by the onClose handler. I think that's correct.

By the way, I've added a comment about that in `core-test/process-associate-request!-test`; perhaps we should document that here.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -281,216 +252,323 @@
                                  :type :broker-unhandled-message)
-                   (i18n/trs "Unhandled message type '{messagetype}' received from '{commonname}' '{remoteaddr}'"))))))
+                   (i18n/trs "Unhandled message type '{messagetype}' received from '{commonname}' '{remoteaddr}'"))
+        connection))))
```
Why does this return connection? It doesn't make changes in this case.

parisiale replied:
Yes, that is wrong.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -281,216 +252,323 @@
+        {:keys [authorized message]} (authorization-check ring-request)
+        allowed (boolean authorized)]
+    (sl/maplog
+      :trace {:messageid    (:id pcp-message)
```
I know I wrote this, but I've noticed we do `(capsule/wrap message)` pretty liberally, so that could be done here as well to simplify the maplog call.

parisiale replied:
Ok. I'll change the signature to directly pass the capsule then use `(capsule/summarize capsule)`.

MikaelSmith replied:
I had a reason that wouldn't work. Be interesting to see what you come up with. Might have been we just needed to create a capsule earlier.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -281,216 +252,323 @@
+   in order, if the message: 1) is an associate-request as expected during
+   Session Association; 2) is authenticated; 3) is authorized; 4) expired"
+  [broker :- Broker capsule :- Capsule connection :- Connection is-association-request :- s/Bool]
+  (if (and (= :open (:state connection)) (not is-association-request))
```
Could this be a `cond` statement instead? Nested ifs are always ugly.

parisiale replied:
Oh, it was screaming for it.

MikaelSmith commented:
Overall this seems to still work with all our clients.

parisiale commented:
Updated.

I don't think we want to update the broker's maps other than in the onClose handler.

I've added the make-data... functions for all the places where the broker creates messages, so to enforce the signatures schemas only during tests. But I want to make sure that we're actually validating the envelope data content of each incoming message, since we dropped all schema checks, so I've added a TODO.

Also, before splitting this in multiple commits, I'd like to fix and merge PCP-524, if you agree that 1) it should be fixed and 2) we need that now.

parisiale commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -281,216 +256,321 @@
+  [broker :- Broker bytes :- message/ByteArray ws :- (s/maybe Websocket)]
+  ;; TODO(ale): since we don't validate responses anymore and the schema
+  ;; checks are not performed in production, ensure that we validate incoming
+  ;; messages (message format and envelope --> in clj-pcp-common)
```
That's what I want to do before moving this to a new PR.

parisiale replied:
Created PCP-525/6/7 for this.

parisiale commented on test/integration/puppetlabs/pcp/broker/service_test.clj:
```diff
@@ -577,3 +608,5 @@
   (with-app-with-config app broker-services no-vnext-config
     (is true)))
+
+(deftest sending-an-invalid-message)
```
That is wrong.

parisiale commented:
Core unit test failing due to missing schema validation; will fix this in PCP-525.

I'm done with this PR; will split changes in different commits.
